Add building the RPM to the make process

pdevine opened the pull request:
This change uses docker-in-docker to build a centos build container
during the build process which then creates the dynamically linked
binary.  This is done because statically linked binaries on Centos/
RHEL/Fedora are a non-starter, and we want the release process to
not be disrupted.  To create the RPMs, you can use the command:

hack/make.sh binary dind-rpm

which will first create the statically linked docker binary which is
then used to run docker-in-docker to run a centos build container.

Commit:
Make a docker-in-docker dynamic binary and add RPM target

This change adds a new docker-in-docker dynamic binary make target which
builds a centos container for creating the dynamically linked binary.
To use it, you first must create the static binary and then call the
dind-dynbinary target.  You can call it like:

$ hack/make.sh binary dind-dynbinary rpm

This would then package the dynamic binary into the rpm after having
created it in the centos build container.  Unfortunately with this approach
you can't create the rpms and the debs with the same command.  They have to
be created separately otherwise the wrong version (static vs. dynamic) gets
packaged.

Signed-off-by: Patrick Devine <patrick.devine@docker.com>

GordonTheTurtle commented:
Can you please sign your commits following these rules:

https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work

The easiest way to do this is to amend the last commit:

~~~console
$ git clone -b "dind-rpm" git@github.com:pdevine/docker.git somewhere
$ cd somewhere
$ git commit --amend -s --no-edit
$ git push -f
~~~
This will update the existing PR, so you do not need to open a new one.

tiborvass commented:
Ping @tianon :)

pdevine commented:
An FYI: In centos:7.1.1503 systemd-container was introduced which breaks compatibility with the systemd-devel which is pulled in by device-mapper-devel.  I'll peg the version of the centos container to 7.0 for now until I have a fix for 7.1

vbatts commented:
is the whole duplication needed though (golang compiler, fpm, registry, etc)? What if you just did the simplest centos DinD build of a dynbinary, and put it on a volume that the existing docker buildtime has access to, then we could just re-use the fpm, golang stuff that is already there.

tiborvass commented:
@pdevine +1 to vbatts

tianon commented:
@vbatts that sounds pretty similar to what's happening in #12111, but for Debian/Ubuntu :+1:

Commit:
Add fixes to the RPM

Fix naming / add some more dependencies

Signed-off-by: Patrick Devine <patrick.devine@docker.com>

Commit:
Various RPM fixes including:
  - Adding missing RPM dependencies.
  - Add sysconfig configuration files to the RPM.
  - Add an epoch to silence the fpm warning.
  - Remove unnecessary empty package.

Signed-off-by: Chad Metcalf <chad@docker.com>

icecrime commented:
Ping @pdevine: can you please take @vbatts comment into account? Also, you might want to squash those commits (there's a merge in there), thanks!

pdevine commented:
@icecrime:  I'll squish down the commits.  If you look at commit b598359, it guts all of the duplication and makes it so you can run:

$ hack/make.sh binary dind-dynbinary rpm ubuntu

which will build both the rpm and deb packages.

icecrime commented:
Thanks! Ping @jfrazelle @tianon @vbatts for review :-)

jfrazelle commented:
SGTM as long as it's ok with @vbatts rpm master, I can make changes to update this to be like @tianons #12111 after we merge that

jfrazelle commented:
or since we haven't depreciated the old way yet, @tianon do you think it is safe to merge this, then I can have a followup pr to make rpm like the new way, then continue down said path

tianon commented:
Even though this is very similar to `hack/make/ubuntu`, I'm still wary of it, especially if we have plans to deprecate it quickly (like we plan to do with `hack/make/ubuntu`).  It's a whole new artifact (and another new way of building said artifact) that we'll be "supporting" even for a short while without any plans to continue doing so.  IMO, if we're going to add RPMs, let's do it right from the beginning instead of settling.

Also, the RPM created here will be specific to `centos:7.0.1406`, and probably won't work properly on CentOS 6 or other RedHat friends unless we get really lucky with `libdevmapper.so` versioning (which is exactly why #12111 builds a separate "build environment" for each distribution version we want to support; that's the only way to ensure we actually link against the appropriate versions of software for that distribution, and is how distributions themselves build packages).

jfrazelle commented:
I opened a new one with fixes here https://github.com/docker/docker/pull/12917
